src/prefect/storage/git.py

```diff
@@ -0,0 +1,149 @@
+import os
+
+from typing import TYPE_CHECKING, Any
+
+from prefect.client import Secret
+from prefect.storage import Storage
+from prefect.utilities.storage import extract_flow_from_file
+from prefect.utilities.git import TemporaryGitRepo
+
+if TYPE_CHECKING:
+    from prefect.core.flow import Flow
+
+
+class Git(Storage):
+    """
+    Git storage class. This class represents the Storage interface for Flows stored
+    in `.py` files in a git repository.
+    This class represents a mapping of flow name to file paths contained in the git repo,
+    meaning that all flow files should be pushed independently. A typical workflow using
+    this storage type might look like the following:
+    - Compose flow `.py` file where flow has Git storage:
+    ```python
+    flow = Flow("my-flow")
+    flow.storage = Git(repo="my/repo", flow_path="/flows/flow.py", repo_host="github.com")
+    ```
+    - Push this `flow.py` file to the `my/repo` repository under `/flows/flow.py`.
+    - Call `prefect register -f flow.py` to register this flow with Git storage.
+    Args:
+        - flow_path (str): A file path pointing to a .py file containing a flow
+        - repo (str): the name of a git repository to store this Flow
+        - repo_host (str, optional): The site hosting the repo. Defaults to 'github.com'
+        - flow_name (str, optional): A specific name of a flow to extract from a file.
+            If not set then the first flow object retrieved from file will be returned.
+        - git_token_secret_name (str, optional): The name of the Prefect Secret containing
+            an access token for the repo. Defaults to None
+        - git_token_username (str, optional): the username associated with git access token,
+            if not provided it will default to repo owner
+        - branch_name (str, optional): branch name, if not specified and `tag` not specified,
+            repo default branch will be used
+        - tag (str, optional): tag name, if not specified and `branch_name` not specified,
+            repo default branch will be used
+        - clone_depth (int): the number of history revisions in cloning, defaults to 1
+        - use_ssh (bool): if True, cloning will use ssh. Ssh keys must be correctly
+            configured in the environment for this to work
+        - format_access_token (bool): if True, the class will attempt to format acess tokens
+            for common git hosting sites
+        - **kwargs (Any, optional): any additional `Storage` initialization options
+    """
+
+    def __init__(
+        self,
+        flow_path: str,
+        repo: str,
+        repo_host: str = "github.com",
+        flow_name: str = None,
+        git_token_secret_name: str = None,
+        git_token_username: str = None,
+        branch_name: str = None,
+        tag: str = None,
```

Comment:
Can I use a commit #?

Reply:
not right now as implemented, I can do some more digging in `dulwich` to see what's possible here

is commit # a common enough pattern that it's must-have?

Reply:
I guess every git tool I've worked with takes branch/commit hash/tag indiscriminately and I expect the same. If it's at all a pain I don't think it should hold this up but I'd make a follow up issue.

Reply:
turned out to be easy enough, just added